### ISSUE TITLE
How to interpret the data?

### ISSUE THREAD

**Issue by personal-coding** [Kostis-S-Z/trading-rl#4]
Is the Test_1 output the hypothetical results?

How do you interpret the reward results (i.e. what does 47587 reward represent)?

Also, my precisions were 0. Any advice why?

Slippage 0.0: 0.00000
Slippage 0.000002: 0.00000
Slippage 0.00002: 0.00000
Precision Long: 0.00 (0 of 2635)
Precision Short: 0.00 (0 of 2558)
Test reward: 47587.05894003946

**Comment by Kostis-S-Z** [comment]
Hello and thanks for the comment,

**TL;DR**: It is hard to give any feedback just from a reward number. The precision metric should not be used as a performance metric of the agent. You get zeros because I had initially removed it from our _trail_ method to prevent confusion but forgot to remove it from writing it in the results.

#### PnL
Firstly, a reminder if you missed it is that, you should check if you are getting this warning: 
"Warning: No method implemented to calculate the PnL! Returning zero..."
Due to certain publication limitations of this project, unfortunately, I am not in the position to publish the backtest method which calculates the PnL metric (that's why the first three lines of "Slippage" are zero) and thus this is something you should add on your own. 

#### Precision
Secondly, the precision metric is more like a heuristic which doesn't necessarily correlate to good financial performance (=high profit). It is based on the idea that if the agent decides that it should go long (buy) in the next step, then this means the exchange rate will either increase or stay the same. Similarly with the short (sell) position. Even though in short sight this might make sense, this is not always the case due to insignificant volatile price changes (=noise). For example, the price might be fluctuating and decrease for a few steps but the agent decides to keeps on holding a long position because it understands it is just noise and that the overall trend will be an increase. Then the precision metric will count those as wrong, even though the agent is following the correct behavior. In conclusion, the precision metric should not be used as a performance metric of the model. That's the reason it doesn't appear on the paper. It was just there in the initial setup of the model to try to provide some additional insight on the behavior of the agent. I removed it as a metric from our _trail_ method (that's why you get zeros in the results) but not from Deng et al's _deng_ (it still works if you use this method) and that's why it still appears in the results. I should probably remove it all together to prevent any further confusion or better document it in the readme, thanks for the note.

_Technical details if you want to experiment_:
The precision metrics are initialized in lines 136 and 137 in trail_env.py but are not used.
```
        self.long_prec = 0
        self.short_prec = 0
```
In deng_env.py you can see the method in lines 190-199 which calculates the precision.
If you want to use this metric anyway, add the `calc_precision` method of deng_env.py to the trail_env.py file and add the lines below before the return of the `get_reward` function. You can also create a pull request with this if you want.
```
pr_val = self.data[self.position - 1] 
self.calc_precision(c_val, pr_val)
```

#### Reward
Concerning the reward, this number by itself doesn't mean much. It is a positive number so that is probably a good sign, but it is also important to compare it to the number of steps per epoch and the number of epochs you trained the model. It would be more useful to take a look at the training plot of the rewards which shows the progress of the agent. If the rewards seem to increase by epochs this probably means you are in the right direction. There is a lot to say about the reward of an RL agent and how it should be interpreted but I think this might be out of the scope of this issue. Feel free to contact me again about this if you are having trouble interpreting your results.

Hope I cleared some confusion and let me know if you have more questions!

**Comment by personal-coding** [comment]
Thanks for the quick response. I had not checked the code thoroughly enough to see the PnL was not there. I will try to backtest on my own, since the trade data is memorized.

**Comment by ScrapeWithYuri** [comment]
I forked your repo and added my EUR-USD dataset (this is hourly vs. 4-hours in your paper). I tested the code:

1) The reward is increasing by epoch
2) The test reward is quite large
3) But, when I try to compare the PnL, the result is a large negative

I grabbed hourly EUR-USD data from 2010-2018, using 2010-2017 as training and 2018 as testing. I've uploaded my results in my fork, data under the Data folder, and raw data under the Dataset folder.

The PnL comparison is under the Dataset folder as well. Could you take a look and see if my logic is wrong regarding the PnL? Essentially, I am looking at the action taken by the agent and assuming the action is taken at the end of the tick (i.e. buy or sell short at the end of the tick).

Thanks,
Yuri
